src/v/cloud_storage/cache_service.cc

```diff
@@ -1288,42 +1289,55 @@ ss::future<> cache::put(
     options.io_priority_class = io_priority;
     auto out = co_await ss::make_file_output_stream(tmp_cache_file, options);
 
-    std::exception_ptr disk_full_error;
+    std::exception_ptr eptr;
+    bool no_space_on_device = false;
     try {
         co_await ss::copy(data, out)
           .then([&out]() { return out.flush(); })
           .finally([&out]() { return out.close(); });
     } catch (const std::filesystem::filesystem_error& e) {
         // For ENOSPC errors, delay handling so that we can do a trim
-        if (e.code() == std::errc::no_space_on_device) {
-            disk_full_error = std::current_exception();
-        } else {
-            throw;
-        }
+        no_space_on_device = e.code() == std::errc::no_space_on_device;
+        eptr = std::current_exception();
+    } catch (...) {
+        // For other errors, delay handling so that we can clean up the tmp file
+        eptr = std::current_exception();
     }
 
-    if (disk_full_error) {
-        vlog(cst_log.error, "Out of space while writing to cache");
+    // If we failed to write to the tmp file, we should delete it, maybe do an
+    // eager trim, and rethrow the exception.
+    if (eptr) {
+        try {
+            co_await delete_file_and_empty_parents(tmp_filepath.native());
+        } catch (...) {
+            vlog(
+              cst_log.error,
+              "Failed to delete tmp file {}: {}",
+              tmp_filepath.native(),
+              std::current_exception());
+        }
 
-        // Block further puts from being attempted until notify_disk_status
-        // reports that there is space available.
-        set_block_puts(true);
+        if (no_space_on_device) {
```

Comment:
Do we need this code path at all given that we're removing the files upon exception? If "no space left" exception is thrown we will try to delete the temporary file and then we will do a trim. But why trimming if we can't leave any tmp files in any circumstances?